robosuite/utils/control_utils.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+import robosuite.utils.transform_utils as trans
+import scipy.linalg
+
+
+def nullspace_torques(mass_matrix, nullspace_matrix, initial_joint, joint_pos, joint_vel):
+    joint_kp = 10
+    joint_kv = np.sqrt(joint_kp) * 2
+    pose_torques = np.dot(mass_matrix, (joint_kp * (
+            initial_joint - joint_pos) - joint_kv * joint_vel))
+    nullspace_torques = np.dot(nullspace_matrix.transpose(), pose_torques)
+    return nullspace_torques
+
+
+def opspace_matrices(mass_matrix, J_full, J_pos, J_ori):
+    mass_matrix_inv = scipy.linalg.inv(mass_matrix)
+
+    # J M^-1 J^T
+    lambda_full_inv = np.dot(
+        np.dot(J_full, mass_matrix_inv),
+        J_full.transpose())
+
+    # (J M^-1 J^T)^-1
+    lambda_full = scipy.linalg.inv(lambda_full_inv)
+
+    # Jx M^-1 Jx^T
+    lambda_pos_inv = np.dot(
+        np.dot(J_pos, mass_matrix_inv),
+        J_pos.transpose())
+
+    # Jr M^-1 Jr^T
+    lambda_ori_inv = np.dot(
+        np.dot(J_ori, mass_matrix_inv),
+        J_ori.transpose())
+
+    # take the inverse, but zero out elements in cases of a singularity
+    # todo: maybe there is an error here? check with danfei and ajay
+    svd_u, svd_s, svd_v = np.linalg.svd(lambda_pos_inv)
+    singularity_threshold = 0.00025
+    svd_s_inv = [0 if x < singularity_threshold else 1. / x for x in svd_s]
+    lambda_pos = svd_v.T.dot(np.diag(svd_s_inv)).dot(svd_u.T)
+
+    svd_u, svd_s, svd_v = np.linalg.svd(lambda_ori_inv)
+    svd_s_inv = [0 if x < singularity_threshold else 1. / x for x in svd_s]
+    lambda_ori = svd_v.T.dot(np.diag(svd_s_inv)).dot(svd_u.T)
+
+    #nullspace
+    Jbar = np.dot(mass_matrix_inv, J_full.transpose()).dot(lambda_full)
+    nullspace_matrix = np.eye(J_full.shape[-1], J_full.shape[-1]) - np.dot(Jbar, J_full)
+
+    return lambda_full, lambda_pos, lambda_ori, nullspace_matrix
+
+
+def orientation_error(desired, current):
+    """
+    Optimized function to determine orientation error from matrices
+    """
+
+    def cross_product(vec1, vec2):
+        mat = np.array(([0, -vec1[2], vec1[1]],
+                        [vec1[2], 0, -vec1[0]],
+                        [-vec1[1], vec1[0], 0]))
+        return np.dot(mat, vec2)
+
+    rc1 = current[0:3, 0]
+    rc2 = current[0:3, 1]
+    rc3 = current[0:3, 2]
+    rd1 = desired[0:3, 0]
+    rd2 = desired[0:3, 1]
+    rd3 = desired[0:3, 2]
+
+    error = 0.5 * (cross_product(rc1, rd1) + cross_product(rc2, rd2) + cross_product(rc3, rd3))
+
+    return error
+
+
+def set_goal_position(delta,
+                       current_position,
+                       position_limit=None,
+                       set_pos=None):
```

Comment:
add docstring - what does this function do?

Reply:
done